File: DataStructure/03.Stack/Java/src/main/java/com/skymoon/RemoveDuplicates.java
package com.skymoon;

import java.util.Deque;
import java.util.LinkedList;

public class RemoveDuplicates {
    public static String removeDuplicates(String S) {
        Deque<Character> stack = new LinkedList<Character>();
        // 遍历字符串
        for (int index=0; index< S.length(); index++){
            if (stack.isEmpty() || S.charAt(index) != stack.peek()){
                stack.push(S.charAt(index));
            }else {
                stack.pop();
            }
        }
        // 栈中元素全部pop出来组成字符串
        String resStr = "";
        while (!stack.isEmpty()){
            resStr = resStr.concat(stack.removeLast().toString());

        }
        return resStr;

    }

    public static void main(String[] args) {
        String a = "abbaca";
        String res = removeDuplicates(a);
        System.out.println(res);

    }
}

File: DataStructure/03.Stack/Go/removeDuplicates.go
package main

func removeDuplicates(S string) string {
	var stack []string
	for i:=0; i< len(S); i++{
		char := S[i]
		if len(stack) == 0 || string(char) != stack[len(stack)-1]{
			stack = append(stack, string(char))
		} else {
			stack = stack[:len(stack)-1]
		}
	}
	res := ""
	for _, ch := range stack{
		res = res + string(ch)
	}
	return res
}

func main() {
	a := "abbaca"
	removeDuplicates(a)
}
